functions/ecs-shutdown-services/src/main/java/com/amazon/aws/partners/saasfactory/saasboost/EcsShutdownServices.java

```diff
@@ -0,0 +1,135 @@
+        }
+        if (Utils.isBlank(API_TRUST_ROLE)) {
+            throw new IllegalStateException("Missing required environment variable API_TRUST_ROLE");
+        }
+        LOGGER.info("Version Info: {}", Utils.version(this.getClass()));
+
+        this.ecs = Utils.sdkClient(EcsClient.builder(), EcsClient.SERVICE_NAME);
+
+        LOGGER.info("Constructor init: {}", System.currentTimeMillis() - startTimeMillis);
+    }
+
+    @Override
+    public Object handleRequest(Map<String, Object> event, Context context) {
+        Utils.logRequestEvent(event);
+
+        ArrayList<Map<String, Object>> provisionedTenants = getProvisionedTenants(context);
+        if (provisionedTenants != null) {
+            LOGGER.info("{} provisioned tenants to process", provisionedTenants.size());
+            for (Map<String, Object> tenant : provisionedTenants) {
+                // The ECS Cluster and Service for each tenant is named with their short id
+                // We could save this info in parameter store with the other tenant infra
+                // pieces so we're not relying on naming convention
+                String cluster = "tenant-" + ((String) tenant.get("id")).substring(0, 8);
+                String service = cluster;
+                Integer count = 0; // Setting the service's desired count to 0 will gracefully remove all running tasks
+
+                try {
+                    DescribeServicesResponse existingServiceSettings = ecs.describeServices(r -> r
+                            .cluster(cluster)
+                            .services(service)
+                    );
+                    for (Service ecsService : existingServiceSettings.services()) {
+                        if (ecsService.desiredCount() > count) {
+                            LOGGER.info("Updating desired count for service " + service + " to " + count);
+                            try {
+                                UpdateServiceResponse updateServiceResponse = ecs.updateService(r -> r
+                                        .cluster(cluster)
+                                        .service(service)
+                                        .desiredCount(count)
+                                );
+                            } catch (SdkServiceException ecsError) {
+                                LOGGER.error("ecs::UpdateService", ecsError);
+                                LOGGER.error(Utils.getFullStackTrace(ecsError));
+                                throw ecsError;
+                            }
+                        }
+                    }
+                } catch (SdkServiceException ecsError) {
+                    LOGGER.error("ecs::DescribeServices", ecsError);
+                    LOGGER.error(Utils.getFullStackTrace(ecsError));
+                    throw ecsError;
+                }
+            }
+        }
+
+        return null;
+    }
+
+    public ArrayList<Map<String, Object>> getProvisionedTenants(Context context) {
```

Comment:
Just an aside, this should really be extrapolated into a TenantsClient object into a layer. The ECSShutdownServices Lambda shouldn't need to know the static path required to get the provisioned tenants.

Reply:
nit: `s/public/private/`

Reply:
Not sure I agree with that -- if I'm going to call the EC2 REST API, I have to know the path to the resource I want to invoke. This isn't an SDK, it's just a REST API.

Reply:
I made it public for unit test visibility (which I didn't get around to yet). Happy to make it protected or package private.

Reply:
I'm simply suggesting that we should eventually write a client abstraction layer to the REST API, so we can safely make changes to the services (server in the client-server model) and versioned changes to the client without breaking existing installations. Each module should only know the direct information it **absolutely has to know**, treating all else as black boxes.